src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -455,7 +455,7 @@ public ReadOnlySequence<T> Slice(SequencePosition start, SequencePosition end)
         public ReadOnlySequence<T> Slice(SequencePosition start)
         {
             BoundsCheck(start);
-            return SliceImpl(start);
+            return SliceImpl(start.GetObject() == null ? Start : start);
```

Comment:
It looks like we are already checking whether start.GetObject() is null in `BoundsCheck`:
```C#
long runningIndex = ((ReadOnlySequenceSegment<T>?)position.GetObject())?.RunningIndex ?? 0;
```

But there isn't a great way to avoid that (and only occurs in one of the branches).

Maybe, store the result of `start.Object == null` in a bool and pass that in to  `BoundsCheck`. Then `runningIndex` can be 0 unless that bool is true (in which case we can get rid of the null checks there). Something to consider, if it can reduce the instruction size of `BoundsCheck` and `Slice`.

```C#
long runningIndex = 0;
if (positionIsNotNull)
{
   Debug.Assert(position.GetObject() != null);
   runningIndex = ((ReadOnlySequenceSegment<Byte>)position.GetObject()).RunningIndex;
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5ilwDcNGsQDMTUgwDCDAN40GipuIA2EAHYBzBgFkAFAGUYARw4x1YGAAUIuAJYY7GhgAdbDp+oCUchUv9qWgxQHOrqdloAkuoAJjAIDAC8DHp6AEow2DEA8uoqAJ5GpuaWRpr45hgAPGz5GDAAfF5u9o4aLADiMBjZwABWMGAYel5eAPwsaaHhUbHxDGNjDHRC1P7+xADswdMRmtFxCKv+AL5+iufKDIHa+kVmFtbubequz55oDMAQECoM9bgMD55Gt1oobjswnsDvNkitLv47FwUgCgQilCCwWCQlDZockil0pkcnlCiYHqUYOVKjU6o1mu92l0ev1BsNRpNdnj4scsWdQWCtpCZvs5kdLvyTkA=)

Reply:
Store `start.GetObject() != null` in a local and re-use in the `SliceImpl` call

Reply:
Right. In the works :)